src/HttpCache/EventListener/ConfigureListener.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace ApiPlatform\Core\HttpCache\EventListener;
+
+use ApiPlatform\Core\Util\RequestAttributesExtractor;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+
+/**
+ * Configures cache HTTP headers for the current response.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+final class ConfigureListener
+{
+    private $maxAge;
+    private $sharedMaxage;
+    private $vary;
+    private $public;
+
+    public function __construct(int $maxAge = null, int $sharedMaxAge = null, array $vary, bool $public = null)
```

Comment:
Let explain with a schema (my last try)...
Here is the actual implementation
![now](https://cloud.githubusercontent.com/assets/578547/24185292/a86a56e0-0ed2-11e7-8a4e-b817a15b9392.png)

My objective is to transform the last response in step **6** into a 304.

* To do it, varnish need a `if-none-match` header in step **5**
* To do it, client need a `etag` header in step **4**
* To do it, varnish needs nothing, it just forward the response from the app from step **3**
* To do it, app must send this header <= this is my point

And Here is the target sequence
![then](https://cloud.githubusercontent.com/assets/578547/24185307/b9d1e8d0-0ed2-11e7-8aaf-c178c0e6f2f1.png)



Your last comment suggest that varnish add this header by it self, with something like that
![other](https://cloud.githubusercontent.com/assets/578547/24185404/52b1a464-0ed3-11e7-8160-f701147b6840.png)

Is it what you mean? Because I doubt varnish add such header by itself (unless you implement this behavior in a VCL which is fine to me too)